src/NServiceBus.Gateway/Gateway.cs

```diff
@@ -50,31 +52,43 @@ protected override void Setup(FeatureConfigurationContext context)
 
             ConfigureTransaction(context);
 
-            string gatewayInputAddress;
-
-            var requiredTransactionSupport = context.Settings.GetRequiredTransactionModeForReceives();
-
-            var gatewayPipeline = context.AddSatellitePipeline("Gateway", requiredTransactionSupport, PushRuntimeSettings.Default, "gateway", out gatewayInputAddress);
-
             var channelManager = CreateChannelManager(context);
 
             Func<string, IChannelSender> channelSenderFactory;
             Func<string, IChannelReceiver> channelReceiverFactory;
             RegisterChannels(context, channelManager, out channelSenderFactory, out channelReceiverFactory);
 
-            gatewayPipeline.Register(ForwardFailedGatewayMessagesToErrorQueueBehavior.StepId, b => new ForwardFailedGatewayMessagesToErrorQueueBehavior(gatewayInputAddress, b.Build<CriticalError>()), "Handles failures when sending messages through the gateway");
+            var gatewayInputAddress = context.Settings.GetTransportAddress(new LogicalAddress(context.Settings.EndpointInstanceName(), "gateway"));
 
-            var retryPolicy = context.Settings.HasSetting("Gateway.Retries.RetryPolicy") ? context.Settings.Get<Func<IncomingMessage, Exception, int ,TimeSpan>>("Gateway.Retries.RetryPolicy") : DefaultRetryPolicy.BuildWithDefaults();
+            var requiredTransactionSupport = context.Settings.GetRequiredTransactionModeForReceives();
 
-            gatewayPipeline.Register(new RetryFailedGatewayMessagesBehavior.Registration(gatewayInputAddress, retryPolicy));
-            gatewayPipeline.Register("GatewaySendProcessor", b => new GatewaySendBehavior(gatewayInputAddress, channelManager, new MessageNotifier(), b.Build<IDispatchMessages>(), context.Settings, CreateForwarder(channelSenderFactory, b.BuildAll<IDataBus>()?.FirstOrDefault()), GetConfigurationBasedSiteRouter(context)), "Processes messages to be sent to the gateway");
-            context.Pipeline.Register("RouteToGateway", b => new RouteToGatewayBehavior(gatewayInputAddress), "Reroutes gateway messages to the gateway");
+            var retryPolicy = context.Settings.GetOrDefault<Func<IncomingMessage, Exception, int, TimeSpan>>("Gateway.Retries.RetryPolicy");
+
+            context.AddSatelliteReceiver("Gateway", gatewayInputAddress, requiredTransactionSupport, PushRuntimeSettings.Default,
+                (config, errorContext) => GatewayRecoverabilityPolicy.Invoke(errorContext, retryPolicy, config),
+                (builder, messageContext) => SendMessage(context, gatewayInputAddress, channelManager, builder, channelSenderFactory, messageContext));
 
+            context.Pipeline.Register("RouteToGateway", b => new RouteToGatewayBehavior(gatewayInputAddress), "Reroutes gateway messages to the gateway");
```

Comment:
The lines below could probably also use the overload which accepts a behavior instance